README
 Net-IRC2 - Client interface to the Internet Relay Chat protocol
=================================================================

INSTALLATION

    perl Build.PL
    ./Build
    ./Build test
    ./Build install


COPYRIGHT AND LICENSE

Copyright (C) 2005 Karl Y. Pradene

This program is released under the following license: GNU General Public License, version 2.


Changes
Revision history for Net-IRC2

0.23    08/20/2005
	Fix version number problem

0.19    08/20/2005
	Try to fix version number problem

0.17    08/20/2005
	Create Net::IRC2::Parser
	add Connection::part
	add Connection::whois

0.13    08/18/2005
	add Event::parent()
	add Connection::parent()

0.11	08/17/2005
	add SIGNATURE file with Module::Signature
	add Connection::notice 
	add Event::from

0.07	08/15/2005
	add Net::IRC2::add_default_handler
	fix Net::IRC2::Event::dump

0.06	08/14/2005
	add Connection::sl                 ( send raw command )
	add Connection::do_one_loop 
	add Event::to                ( alias to Event::Middle )
	fix IRC grammar

0.05    07/30/2005
        
0.02    07/28/2005
        It Work !!!! please try and enjoy. I'am working on doc.
        See DumBot file

0.01    07/27/2005
        First version, released on an unsuspecting world.


lib/Net/IRC2.pm
#
# Copyright 2005, Karl Y. Pradene <knotty@cpan.org> All rights reserved.
#

package Net::IRC2;

use strict;      use warnings   ;
use Exporter                    ;
use Carp                        ;

our @ISA       = qw( Exporter ) ;
our @EXPORT_OK = qw( new      ) ;
our @Export    = qw( new      ) ;

use vars qw( $VERSION $DEBUG)   ;
$VERSION =                       '0.23' ;
$DEBUG   =                            0 ;


sub new         { shift and return bless {@_} } ;

sub newconn     {
    use Net::IRC2::Connection;
    my $self = shift;
    return $self->connections( Net::IRC2::Connection->new( @_,
							   _parent => $self ) );
}

sub add_default_handler { $_[0]->add_handler( [ 'WaterGate' ], $_[1] )          }

sub add_handler         { map { $_->add_handler( @_ ) } @{ shift->connections } }

sub start       {
    use threads;
    # FIXME
    my @threads = map { threads->create( sub { $_->start() } ) } @{$_[0]->connections};
    map { $_->join } @threads ;
}

sub connections {
    my ( $self, $param ) = @_                           ;
    return $self->{'connections'} unless defined $param ;
    push @{$self->{'connections'}}, $param              ;
    return $param                                       ;
                                                        }
sub callback    {
    my ( $self, $param ) = @_                                  ;
    if ( ref $param eq 'CODE' ) {
	map { $_->callback( $param ) } @{ $self->connections } ;
	return 0                                               ;
    }
    return $_[0]->{'callback'}( $param ) if defined $param     ;
                                                               }

1; # End of Net::IRC2



=head1 NAME

Net::IRC2 - Client interface to the Internet Relay Chat protocol.

=head1 VERSION

 !!! UNDER PROGRAMMING !!!
 You can use and feedback is welcome ( in english or french )

=cut

#This is the documentation for the Version __.__.__ of Net::IRC2 , released _______________.

=pod

=head1 SYNOPSIS

 use Net::IRC2                                                        ;
 my $bot  = new Net::IRC2                                             ;
 my $conn = $bot->newconn( uri => 'irc://Nick!User@localhost:6667/' ) ; 
 $conn->mode( $conn->nick, '+B' )                                     ;
 $conn->mode(  '#Ailleurs +m'   )                                     ;
 $bot->add_default_handler( \&process_event )                         ;
 $bot->start                                                          ;
 ...

=head1 DESCRIPTION

This module will provide you an access to the IRC protocol suitable to write your own IRC-Bots, or your
IRC Client. The API will provide you the sames functions than Net::IRC, so change should be trivial.
This module C<use L<Parse::RecDescent>;> by Dr. Conway Damian.

=head1 FUNCTIONS

=over

=item new()

The constructor, takes no argument. Return a Net::IRC2 object. It's your IRC-Bot.

=item newconn()

Make a new connection. Like Net::IRC + can process a home-made tasty pseudo-URI :
irc://Nick!User@localhost:6667/ . Yummy.

=item start()

Start the bot

=item add_handler()

set handler for all messages matching a command in commands list.
 $bot->add_handler( [ '001'..'005' ], \&function ) ;

=item add_default_handler()

The simple way to handle all events with only one function.
set handler for ALL messages
 $bot->add_default_handler( \&function ) ;

=item connections()

return un ARRAY of Net::IRC2::Connection objects

=item irc_grammar()

! Internal !

=item callback()

! DEPRECATED !
 
=back

=head1 AUTHOR

Karl Y. Pradene, C<< <knotty@cpan.org>, irc://knotty@freenode.org/ >> 

=head1 BUGS

Please report any bugs or feature requests to
C<bug-net-irc2@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-IRC2>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SEE ALSO

Perl modules working with IRC connections: Net::IRC, POE::Component::IRC

IRC Request For Comment 1459 L<http://www.ietf.org/rfc/rfc1459.txt?number=1459>

=head1 COPYRIGHT & LICENSE

Copyright 2005, Karl Y. Pradene <knotty@cpan.org> All rights reserved.

This program is released under the following license: GNU General Public License, version 2

This program is free software; you can redistribute it and/or modify it under the terms
of the GNU General Public License version 2 as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program;
if not, write to the 

 Free Software Foundation,
 Inc., 51 Franklin St, Fifth Floor,
 Boston, MA  02110-1301 USA

See L<http://www.fsf.org/licensing/licenses/gpl.html>

=cut


lib/Net/IRC2/Connection.pm
#
# Copyright 2005, Karl Y. Pradene <knotty@cpan.org> All rights reserved.
#
#

package Net::IRC2::Connection   ;

use strict;      use warnings   ;
use Exporter                    ;
use IO::Socket::INET ()         ;
use Net::IRC2::Parser           ;

our @ISA       = qw( Exporter ) ;
our @EXPORT_OK = qw( new      ) ;
our @Export    = qw( new      ) ;

use vars qw( $VERSION $DEBUG )  ;
$VERSION    =                    '0.23' ;
$DEBUG      =                         0 ;


sub new {
    my $class = shift                                                          ;
    my $self = bless { @_ }                                                    ;
    $self->split_uri if exists $self->{'uri'} ;
    my $sock = $self->socket( IO::Socket::INET->new( PeerAddr => $self->server ,
						     PeerPort => $self->port   ,
						     Proto    => 'tcp'         )
			    ) or ( warn "Can't bind : $@\n" and return undef ) ;
    $sock->send( 'PASS ' . $self->pass . "\n"                    .
                 'NICK ' . $self->nick . "\n"                    .
                 'USER ' . $self->user . ' foo.bar.quux '        .
		 $self->server . ' :' . $self->realname . "\n" ) ;
    $self->parser( new Net::IRC2::Parser )                       ;
    return $self                                                 }

sub start { 
    my $self = shift           ;
    1 while $self->do_one_loop }

sub do_one_loop {
    my $self = shift;
    my ( $sock, $parser ) = ( $self->socket, $self->parser );
    my $line = <$sock>;
    my $event = $parser->message( $line ) or warn "Parse error\n$line" and return 0 ;
    if ( $event->command eq 'PING' ) {
	$sock->send( 'PONG ' . $event->trailing. "\n" )                             ;
    }
    $event->{'_parent'} = $self                                                     ;
    $self->chans( scalar $event->trailing ) if $event->command eq 'JOIN'            ;
    if (      defined $self->{ 'callback' }{ $event->command } ) {
	           &{ $self->{ 'callback' }{ $event->command } } ( $self, $event )  ;
    } elsif ( defined $self->{ 'callback' }{   'WaterGate'   } ) {
	           &{ $self->{ 'callback' }{   'WaterGate'   } } ( $self, $event )  ;
    }
    no strict 'refs'                                                                ;
    &{'cb'.$event->command}($self, $event) if defined &{'cb'.$event->command}       ;
    return $event;
}

sub split_uri {
    # http://www.w3.org/Addressing/draft-mirashi-url-irc-01.txt
    # http://www.mozilla.org/projects/rt-messaging/chatzilla/irc-urls.html
    # irc:[<connect-to>[(/<target>[<modifiers>][<query-string>]|<modifiers>)]]
    # http://www.gbiv.com/protocols/uri/rfc/rfc3986.html
    # irc://nick!user@server:port/
    my $self = shift                                                             ;
    if ( exists $self->{'uri'} ) {
	$self->{'uri'} =~ m|^irc://(.+?)!(.+?)@(.+?):(\d+)/|                     ;
	$self->nick($1); $self->user($2); $self->server($3); $self->port($4)     }
                                                                                 }
sub sl        {
    shift->socket->send(    "@_\n" )                                             }


 ##############
# Commands IRC #
 ##############
sub mode    { shift->sl(  'MODE'   . " @_"   ) }
sub join    { shift->sl(  'JOIN'   . " @_"   ) }
sub privmsg { shift->sl( 'PRIVMSG' . " @_"   ) }
sub notice  { shift->sl( 'NOTICE'  . " @_"   ) }
sub part    { shift->sl(  'PART'   . " @_"   ) }
sub whois   { shift->sl(  'WHOIS'  . " @_"   ) }


############
# Accessor #
############
sub nick     { return   $_[0]->{  'Nick'  } = $_[1] || $_[0]->{  'Nick'  } || die 'no nick'    }
sub pass     { return   $_[0]->{'Password'} = $_[1] || $_[0]->{'Password'} || '2 young 2 die'  }
sub port     { return   $_[0]->{  'Port'  } = $_[1] || $_[0]->{  'Port'  } || 6667             }
sub user     { return   $_[0]->{  'user'  } = $_[1] || $_[0]->{  'user'  } || 'void'           }
sub realname { return   $_[0]->{'realname'} = $_[1] || $_[0]->{'realname'} || 'use Net::IRC2'  }
sub server   { return   $_[0]->{ 'Server' } = $_[1] || $_[0]->{ 'Server' } || 'localhost'      }
sub socket   { return   $_[0]->{ 'socket' } = $_[1] || $_[0]->{ 'socket' } || undef            }
sub parser   { return   $_[0]->{ 'parser' } = $_[1] || $_[0]->{ 'parser' }                     }
sub grammar  { return   $_[0]->{'grammar' } = $_[1] || $_[0]->{'grammar' }                     }
sub callback { return   $_[0]->{'callback'} = $_[1]   if ref $_[1] eq 'CODE'                   ;
               return &{$_[0]->{'callback'}}( $_[1] ) if ref $_[1] eq 'Net::IRC2::Events'      }

sub parent   { return   $_[0]->{'_parent' }                                                    }
sub chans    { return push ( @{shift->{'chans'}}, shift )                                      }

sub last_sl  { return   $_[0]->{'last_sl' } = $_[1] || $_[0]->{'last_sl' }                     }

sub add_default_handler { $_[0]->add_handler( [ 'WaterGate' ], $_[1] ) }

sub add_handler { 
    my ( $self, $commands, $callback ) = @_                        ;
    $commands = [ $commands ] unless ref $commands eq 'ARRAY'      ;
    ( map { $self->{'callback'}{$_} = $callback } @$commands )     }

*add_global_handler = \&Net::IRC2::add_handler;

# sub dispatch { }

1;


__END__

=head1 NAME

Net::IRC2::Connection - One connection to an IRC server.

=head1 VERSION

!!! UNDER PROGRAMMING !!! Wait a moment, please hold the line ...

Documentation in progress ...

=head1 FUNCTIONS

=over

=item new()

Make a Connection object. You don't need to make a NET::IRC2 object if
you just want one connection. You should specify nick, server.

 Net::IRC2::Connection::new( nick=>'MyNick', server=>'host.domain.tld' )

=item add_handler()

Add a callback

 $conn->add_handler( 'PRIVMSG', \&callback )
 $conn->add_handler( [ 'PRIVMSG' , 'JOIN' ], \&callback )

=item add_default_handler()

=item start()

Start the client loop

=item do_one_loop()

Process only the next IRC message

=item nick()

Your Nickname

=item user()

=item pass()

The password

=item realname()

=item parent()

return the Net::IRC2 parent object

=item server()

The server like it was specified on creation

=item port()

=item socket()

Return the socket assigned to the connection

=item chans()

=back 

=head2 IRC Commands

=over

=item mode()

=item join()

=item part

=item privmsg()

=item notice()

=item whois

=back

=head1 INTERNALS FUNCTIONS

=over

=item split_uri()

=item grammar()

=item parser()

=item dispatch()

=item callback()

=item sl()

=item last_sl()

=back

=head1 SEE ALSO

Perl modules working with IRC connections: Net::IRC, POE::Component::IRC

IRC Request For Comment 1459 L<http://www.ietf.org/rfc/rfc1459.txt?number=1459>

=head1 COPYRIGHT & LICENSE

Copyright 2005, Karl Y. Pradene <knotty@cpan.org> All rights reserved.

This program is released under the following license: GNU General Public License, version 2

This program is free software; you can redistribute it and/or modify it under the terms
of the GNU General Public License version 2 as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program;
if not, write to the 

 Free Software Foundation,
 Inc., 51 Franklin St, Fifth Floor,
 Boston, MA  02110-1301 USA

See L<http://www.fsf.org/licensing/licenses/gpl.html>

=cut

__END__



lib/Net/IRC2/Event.pm
#
# Copyright 2005, Karl Y. Pradene <knotty@cpan.org> All rights reserved.
#

package Net::IRC2::Event        ;

use strict;      use warnings   ;
use Exporter                    ;

our @ISA       = qw( Exporter ) ;
our @EXPORT_OK = qw( new      ) ;
our @Export    = qw( new      ) ;

use vars qw( $VERSION )         ;
$VERSION =                       '0.23' ;

sub new        { shift and return bless { @_, 'timestamp'=>time }                    }

sub dump       {
    my $self = shift                                                                 ;
    print "------------\n"                                                           .
                                    ' Time     : ' . $self->time       . "\n"        .
                                    ' Orig     : ' . $self->orig                     .
          ( ( $self->prefix     ) ? ' Prefix   : ' . $self->prefix     . "\n" : '' ) .
                                    ' Command  : ' . $self->command    . "\n"        .
          ( ( $self->middle     ) ? ' Middle   : ' . $self->middle     . "\n" : '' ) . 
          ( ( $self->trailing   ) ? ' Trailing : ' . $self->trailing   . "\n" : '' ) .
          ( ( $self->servername ) ? ' Server   : ' . $self->servername . "\n" : '' ) . 
          ( ( $self->nick       ) ? ' Nick     : ' . $self->nick       . "\n" : '' ) . 
          ( ( $self->user       ) ? ' User     : ' . $self->user       . "\n" : '' ) . 
          ( ( $self->host       ) ? ' Host     : ' . $self->host       . "\n" : '' ) .
          ( ( $self->com_str    ) ? ' Com_str  : ' . $self->com_str    . "\n" : '' ) .
          ( ( $self->to         ) ? ' To       : ' . $self->to         . "\n" : '' ) ; 
                                                                                     }
 ##########
# Accessor #
 ##########
sub time       { return $_[0]->{'timestamp' }                                  }
sub orig       { return $_[0]->{   'orig'   }                                  }
sub prefix     { return $_[0]->{  'prefix'  } = $_[1] or $_[0]->{  'prefix'  } }

sub from       { return $_[0]->{   'from'   } = $_[1] || $_[0]->{   'from'   } }
sub servername { return $_[0]->{'servername'} = $_[1] || $_[0]->{'servername'} }
sub nick       { return $_[0]->{   'nick'   } = $_[1] || $_[0]->{   'nick'   } }
sub user       { return $_[0]->{   'user'   } = $_[1] || $_[0]->{   'user'   } }
sub host       { return $_[0]->{   'host'   } = $_[1] || $_[0]->{   'host'   } }

sub command    { return $_[0]->{ 'command'  } = $_[1] || $_[0]->{ 'command'  } }

sub middle     {
    $_[0]->{ 'middle'  } = $_[1] || $_[0]->{ 'middle' }                        ;
    return ( wantarray ) ? $_[0]->{ 'middle' }   : "@{$_[0]->{'middle'}}"      }
sub trailing   {
    $_[0]->{'trailing' } = $_[1] || $_[0]->{'trailing'}                        ;
    return ( wantarray ) ? $_[0]->{'trailing'} : "@{$_[0]->{'trailing'}}"      }


sub com_str    { return $_[0]->{ 'com_str'  } = $_[1] || $_[0]->{ 'com_str'  } }
sub userhost   { warn 'TODO: userhost for '. ref $_[0]                         }

# FIXME !!!
*to = \&middle;

*parent = \&Net::IRC2::Connection::parent;

sub convert {
    my %hash = (
401 => 'NOSUCHNICK'      ,402 => 'NOSUCHSERVER',
403 => 'NOSUCHCHANNEL'   ,404 => 'CANNOTSENDTOCHAN',
405 => 'TOOMANYCHANNELS' ,406 => 'WASNOSUCHNICK',
407 => 'TOOMANYTARGETS'  ,409 => 'NOORIGIN',
411 => 'NORECIPIENT'     ,412 => 'NOTEXTTOSEND',
413 => 'NOTOPLEVEL'      ,414 => 'WILDTOPLEVEL',
421 => 'UNKNOWNCOMMAND'  ,422 => 'NOMOTD',
423 => 'NOADMININFO'     ,424 => 'FILEERROR',
431 => 'NONICKNAMEGIVEN' ,432 => 'ERRONEUSNICKNAME',
433 => 'NICKNAMEINUSE'   ,436 => 'NICKCOLLISION',
441 => 'USERNOTINCHANNEL',442 => 'NOTONCHANNEL',
443 => 'USERONCHANNEL'   ,444 => 'NOLOGIN',
445 => 'SUMMONDISABLED'  ,446 => 'USERSDISABLED',
451 => 'NOTREGISTERED'   ,461 => 'NEEDMOREPARAMS',
462 => 'ALREADYREGISTRED',463 => 'NOPERMFORHOST',
464 => 'PASSWDMISMATCH'  ,465 => 'YOUREBANNEDCREEP',
467 => 'KEYSET'          ,471 => 'CHANNELISFULL',
472 => 'UNKNOWNMODE'     ,473 => 'INVITEONLYCHAN',
474 => 'BANNEDFROMCHAN'  ,475 => 'BADCHANNELKEY',
481 => 'NOPRIVILEGES'    ,482 => 'CHANOPRIVSNEEDED',
483 => 'CANTKILLSERVER'  ,491 => 'NOOPERHOST',
501 => 'UMODEUNKNOWNFLAG',502 => 'USERSDONTMATCH',
300 => 'NONE'            ,302 => 'USERHOST',
303 => 'ISON'            ,301 => 'AWAY',
305 => 'UNAWAY'          ,306 => 'NOWAWAY',
311 => 'WHOISUSER'       ,312 => 'WHOISSERVER',
313 => 'WHOISOPERATOR'   ,317 => 'WHOISIDLE',
318 => 'ENDOFWHOIS'      ,319 => 'WHOISCHANNELS',
314 => 'WHOWASUSER'      ,369 => 'ENDOFWHOWAS',
321 => 'LISTSTART'       ,322 => 'LIST',
323 => 'LISTEND'         ,324 => 'CHANNELMODEIS',
331 => 'NOTOPIC'         ,332 => 'TOPIC',
341 => 'INVITING'        ,342 => 'SUMMONING',
351 => 'VERSION'         ,352 => 'WHOREPLY',
315 => 'ENDOFWHO'        ,353 => 'NAMREPLY',
366 => 'ENDOFNAMES'      ,364 => 'LINKS',
365 => 'ENDOFLINKS'      ,367 => 'BANLIST',
368 => 'ENDOFBANLIST'    ,371 => 'INFO',
374 => 'ENDOFINFO'       ,375 => 'MOTDSTART',
372 => 'MOTD'            ,376 => 'ENDOFMOTD',
381 => 'YOUREOPER'       ,382 => 'REHASHING',
391 => 'TIME'            ,392 => 'USERSSTART',
393 => 'USERS'           ,394 => 'ENDOFUSERS',
395 => 'NOUSERS'         ,200 => 'TRACELINK',
201 => 'TRACECONNECTING' ,202 => 'TRACEHANDSHAKE',
203 => 'TRACEUNKNOWN'    ,204 => 'TRACEOPERATOR',
205 => 'TRACEUSER'       ,206 => 'TRACESERVER',
208 => 'TRACENEWTYPE'    ,261 => 'TRACELOG',
211 => 'STATSLINKINFO'   ,212 => 'STATSCOMMANDS',
213 => 'STATSCLINE'      ,214 => 'STATSNLINE',
215 => 'STATSILINE'      ,216 => 'STATSKLINE',
218 => 'STATSYLINE'      ,219 => 'ENDOFSTATS',
241 => 'STATSLLINE'      ,242 => 'STATSUPTIME',
243 => 'STATSOLINE'      ,244 => 'STATSHLINE',
221 => 'UMODEIS'         ,251 => 'LUSERCLIENT',
252 => 'LUSEROP'         ,253 => 'LUSERUNKNOWN',
254 => 'LUSERCHANNELS'   ,255 => 'LUSERME',
256 => 'ADMINME'         ,257 => 'ADMINLOC1',
258 => 'ADMINLOC2'       ,259 => 'ADMINEMAIL',
209 => 'TRACECLASS'      ,217 => 'STATSQLINE',
231 => 'SERVICEINFO'     ,232 => 'ENDOFSERVICES',
233 => 'SERVICE'         ,234 => 'SERVLIST',
235 => 'SERVLISTEND'     ,316 => 'WHOISCHANOP',
361 => 'KILLDONE'        ,362 => 'CLOSING',
363 => 'CLOSEEND'        ,373 => 'INFOSTART',
384 => 'MYPORTIS'        ,466 => 'YOUWILLBEBANNED',
476 => 'BADCHANMASK'     ,492 => 'NOSERVICEHOST',
);
    foreach ( keys %hash ) {
	$hash{$hash{$_}} = $_;
    }
    return $hash{$_[0]} ;
}

1;

__END__

=head1 NAME

Net::IRC2::Event - A parsed IRC message.

=head1 FUNCTIONS

=over

=item parent()

return the Net::IRC2::Connection parent object

=item dump()

Print a nice formated dump of the Event

=item time()

A timestamp

=item orig()

The original IRC message

=item prefix()

Return the prefix field in IRC message

=item command()

Return the command field in IRC message

=item middle()

Return the midlle field in IRC message

=item trailing()

Return the Trailing field in IRC message

=item com_str()

Should return the command in ALPHA if exist

=item from()

=item servername()

=item nick()

Nickname of sender

=item user()

Username of sender

=item host()

Host of sender

=item userhost()

=item to()

The message destination, could be a chan, you nick ...

the Event's destination

=back

=head1 INTERNALS FUNCTIONS

=over

=item new()

=item convert()

=back

=head1 SEE ALSO

Perl modules working with IRC connections: Net::IRC, POE::Component::IRC

IRC Request For Comment 1459 L<http://www.ietf.org/rfc/rfc1459.txt?number=1459>

=head1 COPYRIGHT & LICENSE

Copyright 2005, Karl Y. Pradene <knotty@cpan.org> All rights reserved.

This program is released under the following license: GNU General Public License, version 2

This program is free software; you can redistribute it and/or modify it under the terms
of the GNU General Public License version 2 as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program;
if not, write to the 

 Free Software Foundation,
 Inc., 51 Franklin St, Fifth Floor,
 Boston, MA  02110-1301 USA

See L<http://www.fsf.org/licensing/licenses/gpl.html>

=cut
